doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -1766,6 +1820,111 @@ The following errors apply to both string and numeral notations:
 
      .. todo note on "single qualified identifiers" https://github.com/coq/coq/pull/11718#discussion_r415076703
 
+.. example:: Number Notation for a non inductive type
+
+   The following example parses and prints :g:`prod unit (prod unit unit)` as :g:`3`.
+   The inductive :g:`I` will be used as :n:`@qualid__ind`.
+
+   .. coqtop:: in
+
+      Inductive I := Iempty : I | Iunit : I | Iprod : I -> I -> I.
+
+   We then define :n:`@qualid__parse` and :n:`@qualid__print`
+
+   .. coqtop:: in
+
+      Definition of_uint (x : Number.uint) : I :=
+        let fix f n := match n with
+          | O => Iempty | S O => Iunit
+          | S n => Iprod Iunit (f n) end in
+        f (Nat.of_num_uint x).
+
+      Definition to_uint (x : I) : Number.uint :=
+        let fix f i := match i with
+          | Iempty => O | Iunit => 1
+          | Iprod i1 i2 => f i1 + f i2 end in
+        Nat.to_num_uint (f x).
+
+      Inductive prod (A : Set) (B : Set) : Set := pair : A -> B -> prod A B.
+
+   the number notation itself
+
+   .. coqtop:: in
+
+      Notation nSet := Set (only parsing).
+      Number Notation nSet via I using (Empty_set Iempty, unit Iunit, prod Iprod)
+        of_uint to_uint : type_scope.
+
+   and check the printer
+
+   .. coqtop:: all
+
+      Local Open Scope type_scope.
+      Check prod unit (prod unit unit).
+
+   and the parser
+
+   .. coqtop:: all
+
+      Set Printing All.
+      Check 3.
+
+.. example:: Number Notation with implicit arguments
+
+   .. coqtop:: all reset
+
+      Require Import Vector.
+      Print Fin.t.
+
+   Note the implicit arguments of :g:`Fin.F1` and :g:`Fin.FS`.
+   They won't appear in the corresponding inductive type.
+
+   .. coqtop:: in
+
+      Inductive I := I1 : I | IS : I -> I.
+
+      Definition of_uint (x : Number.uint) : I :=
+        let fix f n := match n with O => I1 | S n => IS (f n) end in
+        f (Nat.of_num_uint x).
+
+      Definition to_uint (x : I) : Number.uint :=
+        let fix f i := match i with I1 => O | IS n => S (f n) end in
+        Nat.to_num_uint (f x).
+
+      Declare Scope fin_scope.
+      Delimit Scope fin_scope with fin.
+      Local Open Scope fin_scope.
+      Number Notation Fin.t via I using ([Fin.F1] I1, [Fin.FS] IS) of_uint to_uint : fin_scope.
+
+   Now :g:`2` is parsed as :g:`Fin.FS (Fin.FS Fin.F1)`, that is
+   :g:`@Fin.FS _ (@Fin.FS _ (@Fin.F1 _))`.
+
+   .. coqtop:: all
+
+      Check 2.
+
+   .. coqtop:: all
+
+      Check 2 : Fin.t 3.
+
+   .. coqtop:: all fail
+
+      Check 2 : Fin.t 2.
```

Comment:
It's not obvious why the one works and the other doesn't.  Can you explain?  Intuitively, I would expect `Check 2 : Fin.t 2.` would work.  Note that I'm not spending too much time to dive deeply into the feature/description.  OTOH, the doc should be much easier to understand than an academic paper, e.g. don't skip explanations.

Reply:
Indeed, I added more explanations.